### ISSUE TITLE
Services backed by DaemonSet pods should have a hostname based network identifier

### ISSUE THREAD

**Issue by fabiand** [kubernetes/kubernetes#41977]
There should be a way to request that the names of Pods associated to a DaemonSet will carry a pod name which is equal to the hostname.

This should make it easier to address the specific pods using kubedns.

I.e currently a podname os a DS is something like `foo-x15nb` which leads to the fqdn `foox15nb.myservice.mycluster.svc`.

But to make it predictive it would be good if the hostname was used, then it's easier to lookup a psecific service on a specific host: `thefoohost.myservice.mycluster.svc`.

**Comment by kargakis** [comment]
@kubernetes/sig-apps-feature-requests

**Comment by kow3ns** [comment]
Are you asking for the Pod to have the exact same name as the node? If so you could only have one DeamonSet per namespace that has this property without causing a name collision.

**Comment by kargakis** [comment]
It would also block cases like the first point in https://github.com/kubernetes/kubernetes/pull/31693#issuecomment-271339397

**Comment by kargakis** [comment]
> It would also block cases like the first point in #31693 (comment)

OTOH, currently we assume we always run at most one replica per node so this would help in that regard.

**Comment by fabiand** [comment]
@kow3ns you are right, that's borked.

My main point is that there should be away how daemonset pods can be addressed per host through DNS.

Considering that skydns is used, then the scheme is (as you probably know) `$host.$service.$namespace.cluster.local`.
But in the case of daemonsets the `$host` part i sgenerated, thus you can not directly addressed that specific pod, without out-of-band knowledge about the generated name of the pod.
To fix this, a mechanism is required to fixate the hostname for daemonset pods.

**Comment by kargakis** [comment]
Could be dsname-hostname

**Comment by fabiand** [comment]
Where would you use it, @kargakis ?

Eventually a flag on the daemonset spec could help to signal that the real hostname shoul dbe used as a pod hostname, i.e. `hostsHostname: true`.
Once the pod is initiated, the pod's `hostname` field would be set to the hostname.

**Comment by kargakis** [comment]
> Where would you use it, @kargakis ?

Name of the pod.

**Comment by smarterclayton** [comment]
We need to support updates of daemonsets, which would prevent pod name ==
host name.  but pod name == host name + generation might be cool.

**Comment by fabiand** [comment]
Interesting idea.

But thinking about it again - The requirement to habve podname == hostname only arises form the fact that KubeDNS is using the pod hostname

**Comment by fabiand** [comment]
Appologies. Wrong button.

**Comment by smarterclayton** [comment]
Until we guarantee only one pod per node (which I doubt we will do) there's
no way to make podName = hostName and tolerate that.

**Comment by johscheuer** [comment]
Hi,

I tried a naive approach with the following spec:

```yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
spec:
  template:
    spec:
      hostname:
        valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
      containers:
        - name: test
          image: "ubuntu:16.04"
```

but when I try to use this spec I get the following error (even if the fieldpath return a string https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1275: 

```bash
error: error validating "test.yaml": error validating data: expected type string, for field spec.template.spec.hostname, got map; if you choose to ignore these errors, turn validation off with --validate=false
```

Is there a way to tell the downward API to return a string/single value?

**Comment by fabiand** [comment]
Let me rephrase the issue, as it might not be clear from my initial statement.

Whenever a Pod is exposing a service, then the service exposed by a specific pod can be accessed using the fqdn `specifc-pod-name.my-service.my-namespace.svc.cluster.local`.
For regular pods this is okay, as the name `specific-pod-name` is known in advance and can be used to reference to specific pod.
However, in the case of daemon sets the name of a pod is not known in advance, and thus we cna not reference a specific pod.
For daemon sets in particular, the host plays an important role, as daemon sets can be considered to be node local services. Any other pod on a node might be interested to access a specific service on it's own node (for several reasons, one is to avoid network latency by accessing a remote pod providing the same service).

Thus to allow a pod to access a specific service on it's own node, it would be helpful if a daemonset pod can be directly addressed using the service, namespace, and - because it's especially interesting in the daemonset case - the hostname.
I.e.: `<hostname>.<service>.<namespace>.svc.cluster.local`.
With such an address it would be possible for pods to directly access a specific service on a given node.

There could be collisions if a service is bound to more than one pod on a node. I wonder if this could be somehow avoided for the daemonset case.

**Comment by kargakis** [comment]
@fabiand have you looked at the proposal for node-local services?
https://github.com/kubernetes/kubernetes/pull/28637

**Comment by fabiand** [comment]
Would still be important to me.

Might be addressed by #41442
